experiments/Tammy/models/GRU_Tammy.py

```diff
@@ -0,0 +1,189 @@
+        self.update_h2z = nn.Linear(hidden_size, hidden_size, bias=bias)
+
+        # almost output (n)
+        self.output_i2n = nn.Linear(input_size, hidden_size, bias=bias)
+        self.output_h2n = nn.Linear(hidden_size, hidden_size, bias=bias)
+
+        self.init_parameters()
+
+    def reset_gate(self, x: torch.Tensor, h: torch.Tensor) -> torch.Tensor:
+        """
+        x: size (batch_size, input_size)
+        h: size (batch_size, hidden_size)
+        r: size (batch_size, hidden_size)
+        """
+        x_t = self.reset_i2r(x)
+        hs_pre = self.reset_h2r(h)
+        acti = nn.Sigmoid()
+        r = acti(x_t + hs_pre)
+        return r
+
+    def update_gate(self, x: torch.Tensor, h: torch.Tensor) -> torch.Tensor:
+        """
+        x: size (batch_size, input_size)
+        h: size (batch_size, hidden_size)
+        z: size (batch_size, hidden_size)
+        """
+        x_t = self.update_i2z(x)
+        hs_pre = self.update_h2z(h)
+        acti = nn.Sigmoid()
+        z = acti(x_t + hs_pre)
+        return z
+
+    def almost_output(
+        self, x: torch.Tensor, h: torch.Tensor, r: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        x: size (batch_size, input_size)
+        h: size (batch_size, hidden_size)
+        r: size (batch_size, hidden_size)
+        n: size (batch_size, hidden_size)
+        """
+        x_t = self.output_i2n(x)
+        hs_pre = self.output_h2n(h)
+        acti = nn.Tanh()
+        n = acti(x_t + hs_pre)
+        return n
+
+    def forward(self, x: torch.Tensor, h: torch.Tensor = None) -> torch.Tensor:
+        """
+        Computes the forward propagation of the GRU cell
+        Parameters
+        --------
+            input: torch.Tensor
+                Input tensor of shape (batch_size, input_size).
+            hs_pre: torch.Tensor
+                Previous hidden state tensor of shape (batch_size, hidden_size)
+                 Default is None, the initial hidden state is set to zeros.
+        Returns
+        -------
+            hs: torch.Tensor
+                Output hidden state tensor of shape (batch_size, hidden_size).
+        """
+        if h is None:
+            h = torch.zeros(x.size(0), self.hidden_size)
+        r = self.reset_gate(x, h)
+        z = self.update_gate(x, h)
+        n = self.almost_output(x, h, r)
+        hs = (1 - z) * n + z * h
+        return hs
+
+    def init_parameters(self) -> None:
+        """
+        Initialize the weights and biases of the RNN cell
+        followed by Xavier normalization
+        Parameters
+        --------
+            None
+        Returns
+        -------
+            None
+        """
+        for name, param in self.named_parameters():
+            if "weight" in name:
+                torch.nn.init.xavier_uniform_(param)
+            if "bias" in name:
+                param = param.view(1, param.size(0))
+                torch.nn.init.xavier_uniform_(param)
+
+
+class GRU(nn.Module):
```

Comment:
This supports many-to-many with T_x = T_y only (https://stanford.edu/~shervine/teaching/cs-230/cheatsheet-recurrent-neural-networks)